contracts/RelayHub.sol

```diff
@@ -231,12 +239,12 @@ contract RelayHub is RelayHubApi {
         bytes memory transaction = abi.encodePacked(encoded_function,from);
         bool success;
         bool success_post;
-
+        uint balance_before = balances[to];
         (success, ) = to.call.gas(gas_limit)(transaction); // transaction must end with @from at this point
-        nonces[from]++;
         transaction = abi.encodeWithSelector(RelayRecipient(to).post_relayed_call.selector, relay_addr, from, encoded_function, success, (gas_overhead+initial_gas-gasleft()), transaction_fee);
         (success_post, ) = to.call.gas((gas_overhead+initial_gas-gasleft()))(transaction);
         require(success_post, "post_relayed_call reverted - reverting the relayed transaction");
+        require(balance_before <= balances[to], "Moving funds during relayed transaction disallowed");
```

Comment:
- did you really mean `RelayHub.balances`, or relaying a "real" `payable` method?
- in any case, it doesn't matter: we *do* allow moving funds (e.g. relaying erc20's **transfer**), and `relay()` method is not payable anyway.

Reply:
What we disallow is moving deposits made to RelayHub by the recipient during a relayed transaction, otherwise it could lead to the following attack:
A recipient with enough balance accepts a relayed transaction, the relay sends it through the hub, and inside the relayed function call the recipient withdraws its balance from the hub, so that the relay will get reverted on `require(balances[to] >= charge, "insufficient funds");`.
In order to prevent that, we require that the recipient's balance is not lower than it was before the relayed function call.